src/entities/priceImpact/index.ts

```diff
@@ -45,4 +54,134 @@ export class PriceImpact {
+        const remainingDiffIndex = await zeroOutDiffs(diffs, diffBPTs);
+
+        // get relevant amounts for price impact calculation
+        const amountInitial = parseFloat(
+            formatUnits(
+                amountsIn[remainingDiffIndex].amount,
+                amountsIn[remainingDiffIndex].token.decimals,
+            ),
+        );
+        const amountDiff = parseFloat(
+            formatUnits(
+                abs(diffs[remainingDiffIndex]),
+                amountsIn[remainingDiffIndex].token.decimals,
+            ),
+        );
+
+        // calculate price impact using ABA method
+        const priceImpact = amountDiff / amountInitial / 2;
+        return PriceImpactAmount.fromDecimal(`${priceImpact}`);
+
+        // helper functions
+
+        async function zeroOutDiffs(diffs: bigint[], diffBPTs: bigint[]) {
+            let minNegativeDiffIndex = 0;
+            const nonZeroDiffs = diffs.filter((d) => d !== 0n);
+            for (let i = 0; i < nonZeroDiffs.length - 1; i++) {
+                const minPositiveDiffIndex = diffBPTs.findIndex(
+                    (diffBPT) =>
+                        diffBPT === min(diffBPTs.filter((a) => a > 0n)),
+                );
+                minNegativeDiffIndex = diffBPTs.findIndex(
+                    (diffBPT) =>
+                        diffBPT === max(diffBPTs.filter((a) => a < 0n)),
+                );
+
+                let kind: SwapKind;
+                let givenIndex: number;
+                let resultIndex: number;
+                if (
+                    diffBPTs[minPositiveDiffIndex] <
+                    abs(diffBPTs[minNegativeDiffIndex])
+                ) {
+                    kind = SwapKind.GivenIn;
+                    givenIndex = minPositiveDiffIndex;
+                    resultIndex = minNegativeDiffIndex;
+                } else {
+                    kind = SwapKind.GivenOut;
+                    givenIndex = minNegativeDiffIndex;
+                    resultIndex = minPositiveDiffIndex;
+                }
+
+                const resultAmount = await doQuerySwap({
+                    poolId: poolState.id,
+                    kind,
+                    tokenIn: poolTokens[minPositiveDiffIndex].toInputToken(),
+                    tokenOut: poolTokens[minNegativeDiffIndex].toInputToken(),
+                    givenAmount: abs(diffs[givenIndex]),
+                    rpcUrl: input.rpcUrl,
+                    chainId: input.chainId,
+                });
+
+                diffs[givenIndex] = 0n;
+                diffBPTs[givenIndex] = 0n;
+                diffs[resultIndex] = diffs[resultIndex] + resultAmount.amount;
+                diffBPTs[resultIndex] = await queryBPTForDiffAtIndex(
+                    resultIndex,
+                );
+            }
+            return minNegativeDiffIndex;
+        }
+
+        async function queryBPTForDiffAtIndex(i: number): Promise<bigint> {
```

Comment:
Maybe would worth to have a comment for this function, the name is not very explicative, but I understood that this function queries the bptOut amount for an amount delta (amountOut - amountIn), given the token index, right?
Obs.: What do you think of the suggested name: "queryBptOutForSingleTokenDiff(tokenIndex: number)"?

Reply:
Yeah, I agree this function name could be improved. We should likely make it more explicit that is an `addLiquidity` query as well.
Let me know what you think of this suggestion:
```suggestion
        async function queryAddLiquidityForTokenDelta(tokenIndex: number): Promise<bigint> {
```